nussl/separation/duet.py

```diff
@@ -0,0 +1,440 @@
+    def _smooth_matrix(matrix, kernel):
+        """Performs two-dimensional convolution in order to smooth the values of matrix elements.
+
+        (similar to low-pass filtering)
+
+        Parameters:
+            matrix (np.array): a 2D Numpy matrix to be smoothed
+            kernel (np.array): a 2D Numpy matrix containing kernel values
+        Note:
+            if Kernel is of size 1 by 1 (scalar), a Kernel by Kernel matrix of 1/Kernel**2 will be used as the matrix
+            averaging kernel
+        Output:
+            smoothed_matrix (np.array): a 2D Numpy matrix containing a smoothed version of Mat (same size as Mat)
+        """
+
+        # check the dimensions of the Kernel matrix and set the values of the averaging
+        # matrix, kernel_matrix
+        if np.prod(kernel.shape) == 1:
+            kernel_matrix = np.ones((kernel, kernel)) / kernel ** 2
+        else:
+            kernel_matrix = kernel
+
+        # make kernel_matrix have odd dimensions
+        krow, kcol = np.shape(kernel_matrix)
+        if np.mod(krow, 2) == 0:
+            kernel_matrix = signal.convolve2d(kernel_matrix, np.ones((2, 1))) / 2
+            krow += 1
+
+        if np.mod(kcol, 2) == 0:
+            kernel_matrix = signal.convolve2d(kernel_matrix, np.ones((1, 2))) / 2
+            kcol += 1
+
+        # adjust the matrix dimension for convolution
+        copy_row = int(np.floor(krow / 2))  # number of rows to copy on top and bottom
+        copy_col = int(np.floor(kcol / 2))  # number of columns to copy on either side
+
+        # TODO: This is very ugly. Make this readable.
+        # form the augmented matrix (rows and columns added to top, bottom, and sides)
+        matrix = np.mat(matrix)  # make sure Mat is a Numpy matrix
+        augmented_matrix = np.vstack(
+            [
+                np.hstack(
+                    [matrix[0, 0] * np.ones((copy_row, copy_col)),
+                     np.ones((copy_row, 1)) * matrix[0, :],
+                     matrix[0, -1] * np.ones((copy_row, copy_col))
+                     ]),
+                np.hstack(
+                    [matrix[:, 0] * np.ones((1, copy_col)),
+                     matrix,
+                     matrix[:, -1] * np.ones((1, copy_col))]),
+                np.hstack(
+                    [matrix[-1, 1] * np.ones((copy_row, copy_col)),
+                     np.ones((copy_row, 1)) * matrix[-1, :],
+                     matrix[-1, -1] * np.ones((copy_row, copy_col))
+                     ]
+                )
+            ]
+        )
+
+        # perform two-dimensional convolution between the input matrix and the kernel
+        smooted_matrix = signal.convolve2d(augmented_matrix, kernel_matrix[::-1, ::-1], mode='valid')
+
+        return smooted_matrix
+
+    def make_audio_signals(self):
+        """Returns the extracted signals
+
+        Returns:
+            signals (List[AudioSignal]): List of AudioSignals extracted using DUET.
+
+        """
+        signals = []
```

Comment:
Because this is a public function, check to make sure that everything you need is set (correctly) before you do anything. So that's probably just making sure that all of the attributes are not `None`. You can probably just get away with checking to make sure that `len(self.masks) > 0` because the only way that the mask array is filled is if you ran the algorithm. If it fails, throw an exception.